File: minHeap.cpp
#include <iostream>
#include <string>
#include <vector>
using namespace std;
class MinHeap {
public:
	int Capacity;
	vector<int> storage;
	int size = 0;
	MinHeap(int Capacity) {
		this->Capacity = Capacity;
		storage.resize(Capacity);
	}

	double getParentIndex(int index) {
		return (index - 1) / 2.0;
	}
	int getLeftChildIndex(int index) {
		return 2 * index + 1;
	}
	int getRightChildIndex(int index) {
		return 2 * index + 2;
	}
	bool hasParent(int index) {
		return getParentIndex(index) >= 0;
	}
	bool hasLeftChild(int index) {
		return getLeftChildIndex(index) < size;
	}
	bool hasRightChild(int index) {
		return getRightChildIndex(index) > size;
	}
	bool isFull() {
		return storage.size() == size;
	}
	void swap(int index1, int index2, vector<int> &li) {
		int temp = li[index1];
		li[index1] = li[index2];
		li[index2] = temp;

	}
	void insert(int data) {
		if (isFull()) {
			cout << "heap is full" << endl;
		}
		storage[size] = data;
		size += 1;
		heapifyUp();
	}
	void recursive_insert(int data) {
		if (isFull()) {
			cout << "heap is full" << endl;
		}
		storage[size] = data;
		size += 1;
		recursive_heapifyUp(size-1);
	}
	void recursive_heapifyUp(int index) {
		if (hasParent(index) and storage[getParentIndex(index)] > storage[index]) {
			swap(getParentIndex(index), index, storage);
			recursive_heapifyUp(index);
		}
	}
	void heapifyUp() {
		int index = size - 1;
		while (hasParent(index) and storage[getParentIndex(index)] > storage[index]) {
			swap(getParentIndex(index), index, storage);
		}
	}
	int remove() {
		int data = storage[0];
		storage[0] = storage[size - 1];
		storage[size - 1] = 0;
		size -= 1;
		heapifyDown();
		return data;

	}
	int recursive_remove() {
		int data = storage[0];
		storage[0] = storage[size - 1];
		storage[size - 1] = 0;
		size -= 1;
		recursive_heapifyDown(0);
		return data;

	}
	void recursive_heapifyDown(int index) {
		if (hasLeftChild(index)) {
			int smallerIndex = getLeftChildIndex(index);
			if (hasRightChild(index) and storage[getRightChildIndex(index)] < storage[smallerIndex]) {
				smallerIndex = getRightChildIndex(index);
			}
			if (storage[index] < storage[smallerIndex]) {
				return;
			}
			else
				swap(index, smallerIndex, storage);
			recursive_heapifyDown(smallerIndex);
		}
	}
	void heapifyDown() {
		int index = 0;
		while (hasLeftChild(index)) {
			int smallerIndex = getLeftChildIndex(index);
			if (hasRightChild(index) and storage[getRightChildIndex(index)] < storage[smallerIndex]) {
				smallerIndex = getRightChildIndex(index);
			}
			if (storage[index] < storage[smallerIndex]) {
				break;
			}
			else
				swap(index, smallerIndex, storage);
			index = smallerIndex;
		}
	}
};
int main() {
	MinHeap heap(5);
	heap.recursive_insert(5);
	heap.recursive_insert(3);
	heap.recursive_insert(6);
	heap.recursive_remove();
	for (int i = 0; i < heap.storage.size(); i++) {
		cout << heap.storage[i] << '\t';
	}
	
	
	return 0;
}

File: minHeap.py
class MinHeap:
    def __init__(self, capacity):
        self.storage = [0] * capacity
        self.capacity = capacity
        self.size = 0

    def getParentInd(self, index):
        return (index - 1) // 2

    def getLeftChildInd(self, index):
        return 2*index + 1

    def getRightChildInd(self, index):
        return 2 * index + 2

    def hasParent(self, index):
        return self.getParentInd(index) >= 0

    def hasLeftChild(self, index):
        return self.getLeftChildInd(index) < self.size

    def hasRightChild(self, index):
        return self.getRightChildInd(index) < self.size

    def isFull(self):
        return self.size == self.capacity

    def swap(self, index1, index2):
        temp = self.storage[index1]
        self.storage[index1] = self.storage[index2]
        self.storage[index2] = temp

    def insert(self, data):
        if self.isFull():
            print("heap is full")
        self.storage[self.size] = data
        self.size += 1
        self.heapifyUp()
    def recursive_insert(self, data):
        if self.isFull():
            print("heap is full")
        self.storage[self.size] = data
        self.size += 1
        self.recursiveHeapifyUp(self.size - 1)

    def heapifyUp(self):
        index = self.size - 1
        while self.hasParent(index) and self.storage[self.getParentInd(index)] > self.storage[index]:
            self.swap(self.getParentInd(index), index)
            index = self.getParentInd(index)

    def recursiveHeapifyUp(self, index):
        if self.hasParent(index) and self.storage[self.getParentInd(index)] > self.storage[index]:
            self.swap(self.getParentInd(index), index)
            self.recursiveHeapifyUp(self.getParentInd(index))

    def remove(self):
        if self.size == 0:
            print("heap is empty")
        data = self.storage[0]
        self.storage[0] = self.storage[self.size - 1]
        self.storage[self.size - 1] = 0
        self.size -= 1
        self.heapifyDown()
        return data

    def recursive_remove(self):
        if self.size == 0:
            print("heap is empty")
        data = self.storage[0]
        self.storage[0] = self.storage[self.size - 1]
        self.storage[self.size - 1] = 0
        self.size -= 1
        self.recursive_heapifyDown(0)
        return data
    def heapifyDown(self):
        index = 0
        while self.hasLeftChild(index):
            smallerChildIndex = self.getLeftChildInd(index)
            if self.hasRightChild(index) and self.storage[self.getRightChildInd(index)] < self.storage[smallerChildIndex]:
                smallerChildIndex = self.getRightChildInd(index)
            if self.storage[index] < self.storage[smallerChildIndex]:
                break
            else:
                self.swap(index, smallerChildIndex)
            index = smallerChildIndex

    def recursive_heapifyDown(self, index):
        if self.hasLeftChild(index):
            smallerChildIndex = self.getLeftChildInd(index)
            if self.hasRightChild(index) and self.storage[self.getRightChildInd(index)] < self.storage[smallerChildIndex]:
                smallerChildIndex = self.getRightChildInd(index)
            if self.storage[index] < self.storage[smallerChildIndex]:
                return
            else:
                self.swap(index, smallerChildIndex)
            index = smallerChildIndex
            self.recursive_heapifyDown(index)


ob = MinHeap(5)
ob.insert(5)
ob.insert(3)
ob.insert(6)
print(ob.storage)